### ISSUE TITLE
[pickers] Add support for date-fns 3.0

### ISSUE THREAD

**Issue by Studio384** [mui/mui-x#11454]
### Steps to reproduce

Install `date-fns` 3.0.0 or higher when using `@mui/x-date-pickers`.

### Current behavior

Currently, when using @mui/x-date-pickers together with date-fns version 3.0.0, npm will return the following error:

```
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: --
npm WARN Found: date-fns@2.30.0
npm WARN node_modules/date-fns
npm WARN   date-fns@"3.0.0" from the root project
npm WARN   1 more (@mui/x-date-pickers)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peerOptional date-fns@"^2.25.0" from @mui/x-date-pickers@6.18.5
npm WARN node_modules/@mui/x-date-pickers
npm WARN   @mui/x-date-pickers@"6.18.5" from the root project
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '--',
npm WARN EBADENGINE   required: { node: '18.x', npm: '8.x' },
npm WARN EBADENGINE   current: { node: 'v20.9.0', npm: '10.2.4' }
npm WARN EBADENGINE }
```

The application won't run.

### Expected behavior

The application does run and installing date-fns version 3.0 works without errors.

**Search keywords**: date time picker

**Comment by michelengelen** [comment]
Hey @Studio384 
We are currently only supporting `date-fns` between version `2.25.0` and `2.30.0` which is indicated by this line in the `package.json`
https://github.com/mui/mui-x/blob/58246d3b5fc30bdc5adda895f31888d0401eac4a/packages/x-date-pickers-pro/package.json#L59

I am not sure if we are planning to support `date-fns@3.0.0` with the new major.
Are there plans to adopt the new major of `date-fns` @LukasTy?

**Comment by flaviendelangle** [comment]
`date-fns@3.0.0` was released yesterday so we did not discuss it yet tbh :laughing: 

Looks like we won't be able to support both v2 and v3 because they changes the import strategy.
cc @LukasTy

**Comment by LukasTy** [comment]
> Looks like we won't be able to support both v2 and v3 because they changes the import strategy.

Have you tested it?
I'm going through the blog post and it seems that they have a fallback of exporting the default. 🤔

**Comment by michelengelen** [comment]
> `date-fns@3.0.0` was released yesterday so we did not discuss it yet tbh 😆

Oh, I did not check the release date ... I suspected it to be fairly new, since it's just the `3.0.0` version available! :D

**Comment by flaviendelangle** [comment]
>  it seems that they have a fallback of exporting the default.

I spoke too fast then sorry

**Comment by flaviendelangle** [comment]
:grimacing: We may need to ask `date-fns` maintainer a was to access this variable

**Comment by jinsley8** [comment]
I have a similar issue related to the `date-fns` v3 import changes.

```
./node_modules/@mui/x-date-pickers/AdapterDateFns/AdapterDateFns.js:59:0

Module not found: Package path ./_lib/format/longFormatters is not exported from package /app/node_modules/date-fns (see exports field in /app/node_modules/date-fns/package.json)
```

**Comment by LukasTy** [comment]
Hello.
If anyone is eager to experiment with date-fns v3 and is already on v7 alpha of `@mui/x-date-pickers`, feel free to try out this build: `https://pkg.csb.dev/mui/mui-x/commit/b12bff5a/@mui/x-date-pickers` (put it instead of your current package version):
```diff
- "@mui/x-date-pickers": "^7.0.0-alpha.6",
+ "@mui/x-date-pickers": "https://pkg.csb.dev/mui/mui-x/commit/b12bff5a/@mui/x-date-pickers"
```
You can find usage instructions [here](https://deploy-preview-11462--material-ui-x.netlify.app/x/react-date-pickers/adapters-locale/#with-date-fns).

**Comment by Nefcanto** [comment]
For me, I get this error. I run my code inside a docker container:

```
DockerContainer | /Infra/Infra/node_modules/esbuild/lib/main.js:1651
DockerContainer |   let error = new Error(text);
DockerContainer |               ^
DockerContainer | 
DockerContainer | Error: Build failed with 1 error:
DockerContainer | node_modules/@mui/x-date-pickers/AdapterDateFns/AdapterDateFns.js:59:27: ERROR: [plugin: vite:dep-pre-bundle] Missing "./_lib/format/longFormatters" specifier in "date-fns" package
DockerContainer |     at failureErrorWithLog (/Infra/Infra/node_modules/esbuild/lib/main.js:1651:15)
DockerContainer |     at /Infra/Infra/node_modules/esbuild/lib/main.js:1059:25
DockerContainer |     at /Infra/Infra/node_modules/esbuild/lib/main.js:1527:9
DockerContainer |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
DockerContainer |   errors: [Getter/Setter],
DockerContainer |   warnings: [Getter/Setter]
DockerContainer | }
DockerContainer | 
DockerContainer | Node.js v20.9.0

```

**Comment by LukasTy** [comment]
> For me, I get this error. I run my code inside a docker container:

Which version of `date-fns` are you using? 🤔 
Did you try the latest (at the moment) `v3.0.6`?

**Comment by ruiaraujo012** [comment]
Same here with the latest version of `date-fns`

**Comment by Nefcanto** [comment]
@LukasTy, I'm using `3.0.6`. The latest now.

**Comment by aemonm** [comment]
@LukasTy  does the alpha version (`"https://pkg.csb.dev/mui/mui-x/commit/b12bff5a/@mui/x-date-pickers"`) address the issue with the `longFormatters` export not being found by mui?

**Comment by LukasTy** [comment]
> does the alpha version (`"https://pkg.csb.dev/mui/mui-x/commit/b12bff5a/@mui/x-date-pickers"`) address the issue with the `longFormatters` export not being found by mui?

Depends. It does seem to work in our codesandboxes (using react-create-app), but using it within a vite + next app produces the mentioned error.
We'll investigate the possible solutions for that specific problem. 👍

**Comment by LukasTy** [comment]
Getting back to confirm that there does not seem to be any workaround except for basically copy-pasting the logic in `date-fns/_lib/format/longFormatters`.
Will ask if they'd be willing to export it or we'll need to look for other options. 🤔

**Comment by michelengelen** [comment]
The opening question seems to have been tackled. Can we close this now @LukasTy?
If this issue is really pressing for you @aemonm feel free to open a separate ticket so we can track this more efficiently! 🙇🏼

**Comment by LukasTy** [comment]
No, I don't think that the problem has been tackled, because we still can not successfully support date-fns@v3 at this moment, even with the new Adapter there are importing problems depending on the runtime environment.

**Comment by vishal-kadmos** [comment]
Observed same issue while updating date-fns to v3.1.0. Looking forward for a fix from community.

**Comment by valourus** [comment]
I would also love to see a update to ^3.0.0

**Comment by flaviendelangle** [comment]
We are in good shape to have an update pretty soon :ok_hand:

**Comment by LukasTy** [comment]
The support for `date-fns@v3` has been released in versions `6.19.0` and `7.0.0-alpha.8`.
Feel free to change the import:
```diff
- import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
+ import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
```
and use at least version `3.2.0` of the `date-fns` package.

**Comment by PzYon** [comment]
Thanks for the update. I tried `AdapterDateFnsV3` with following dependencies:
```
"@mui/x-date-pickers": "6.19.0",
"date-fns": "3.2.0",
```
and following component:
```
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFnsV3";
import de from "date-fns/locale/de";

const DemoComponent: React.FC = () => {
  return (
    <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={de}>
    </LocalizationProvider>
  );
};
```

This now leads to following error:

`Type 'typeof import("D:/Repos/engraved/app/node_modules/date-fns/locale/de")' is missing the following properties from type 'Locale': code, formatDistance, formatRelative, localize, and 2 more.
`

**Comment by LukasTy** [comment]
@PzYon Please update the `date-fns/locale` import as well.
`date-fns` has changed all exports to named ones instead of default in v3.
You can find more information about different usage here: https://next.mui.com/x/react-date-pickers/adapters-locale/#with-date-fns

**Comment by PzYon** [comment]
I'm sorry, I missed that. Now it works, thank you very much!

```TypeScript
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFnsV3";
import { de } from "date-fns/locale/de";

export const DemoComponent: React.FC = () => {
  return (
    <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={de}>
    </LocalizationProvider>
  );
};
```

**Comment by Nefcanto** [comment]
@LukasTy, I updated to version `6.19.0` and it worked without changing anything in my code. How could it be?

These are my imports in the `main.jsx`:

```
import { StrictMode } from 'react'
import { createRoot } from 'react-dom/client'
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns'
import { AdapterDateFnsJalali } from '@mui/x-date-pickers/AdapterDateFnsJalali'
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider'
import { ThemeProvider } from '@mui/material/styles'
import { createTheme } from '@mui/material/styles'
import { BrowserRouter } from 'react-router-dom'
import rtlPlugin from 'stylis-plugin-rtl'
import { CacheProvider } from '@emotion/react'
import createCache from '@emotion/cache'
import { prefixer } from 'stylis'
```

**Comment by LukasTy** [comment]
@Nefcanto Did you also update to `date-fns` v3?
If not, then you should not need to make any changes. 👌

Question out of curiosity: what is the reason that you are importing the jalali adapter besides the regular one? 🤔

**Comment by Nefcanto** [comment]
@LukasTy, my bad. The `node_modules` was linked to another directory and when I used `npm list date-fns` it found a mismatch. I had to change it to `@mui/x-date-pickers/AdapterDateFnsV3` as you mentioned.

We use jalali to support jalali calendar.

**Comment by rwb196884** [comment]
Still broken
```
[commonjs--resolver] Missing "./_lib/format/longFormatters" specifier in "date-fns" package
error during build:
Error: Missing "./_lib/format/longFormatters" specifier in "date-fns" package
    at e (file:///C:/Work/Azure/zzz/client/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:21445:25)
    at n (file:///C:/Work/Azure/zzz/client/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:21445:627)
    at o (file:///C:/Work/Azure/zzz/client/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:21445:1297)
    at resolveExportsOrImports (file:///C:/Work/Azure/zzz/client/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:28746:20)
    at resolveDeepImport (file:///C:/Work/Azure/zzz/client/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:28765:31)
    at tryNodeResolve (file:///C:/Work/Azure/zzz/client/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:28453:20)
    at Object.resolveId (file:///C:/Work/Azure/zzz/client/node_modules/vite/dist/node/chunks/dep-bb8a8339.js:28212:28)
    at file:///C:/Work/Azure/zzz/client/node_modules/rollup/dist/es/shared/node-entry.js:25544:40
    at async PluginDriver.hookFirstAndGetPlugin (file:///C:/Work/Azure/zzz/client/node_modules/rollup/dist/es/shared/node-entry.js:25444:28)
    at async resolveId (file:///C:/Work/Azure/zzz/client/node_modules/rollup/dist/es/shared/node-entry.js:24117:26)
```

Last working version is 2.30.0.

**Comment by LukasTy** [comment]
> Still broken

@rwb196884 Which Adapter are you using?
Have you checked [documentation](https://mui.com/x/react-date-pickers/adapters-locale/#with-date-fns) or discussion above?

**Comment by allicanseenow** [comment]
Changing from AdapterDateFns to AdapterDateFnsV3 does fix the issue.

**Comment by mahechahe** [comment]
> Cambiar de AdapterDateFns a AdapterDateFnsV3 soluciona el problema.

Esto funciono para mi

**Comment by harbhub** [comment]
I have the same issue that some others have mentioned.  My only hunch is that it might be related to how Vite does its pre bundling.

**Relevant libraries:**
vite ^5.4.3
@mui/material ^6.0.2
@mui/x-date-pickers ^7.16.0
date-fns ^3.6.0

**Everything works properly if I use date-fns@v2.30.0** and import the old AdapterDateFns as such:
`import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns'`

However, **I need to upgrade to the latest date-fns@v3.6.0** which causes my code to break even if I don't use the AdapterDateFns anywhere in the code.  It doesn't matter if I replace AdapterDateFns entirely by using AdapterDayjs instead or by replacing it with the AdapterDateFnsV3 version of the adapter.  **Regardless of what I do, it somehow throws the same error when trying to run the app via vite.**

**Here are some of the error details:**
[ERROR] Missing "./_lib/format/longFormatters" specifier in "date-fns" package [plugin vite:dep-pre-bundle]
 The plugin "vite:dep-pre-bundle" was triggered by this import
    node_modules/@mui/x-date-pickers/AdapterDateFns/AdapterDateFns.js:46:27:
      46 │ import longFormatters from 'date-fns/_lib/format/longFormatters';

**Comment by harbhub** [comment]
I fixed it by updating the Material UI library in this manner:

.../node_modules/@mui/x-date-pickers/AdapterDateFns/AdapterDateFns.js
// import longFormatters from 'date-fns/_lib/format/longFormatters';
import { longFormatters } from 'date-fns';

I commented out the old import and replaced it with the proper one based on how date-fns@v3.6.0 works.

I'm not sure why this function (AdapterDateFns) is even relevant since I'm not using it anywhere.  Instead, I'm using AdapterDateFnsV3 as advised by the MUI docs.  For whatever reason, I had to simply fix the source code of the @mui/x-date-pickers to properly import the longFormatters even though I don't use AdapterDateFns to begin with (as I use AdapterDateFnsV3 instead).

Leaving this solution here in case it happens to help anyone else.  I would appreciate an explanation if anyone happens to have one.  My suspicion is that the issue is related to Vite and its pre-bundle functionality.

**Comment by LukasTy** [comment]
@harbhub would you be able to provide a minimal reproduction example so that we can look into this? 🤔 
It could be the behavior of ESBuild or another tool that Vite uses under the hood for dependency parsing.
If this is a clear issue, it would be great to at least include this "gotcha" in the FAQ. 🙈 

You could also create a separate issue for this to avoid cluttering discussions here. 😉

**Comment by harbhub** [comment]
@LukasTy I think just sharing the "solution" to my problem here will suffice.  It would be hard to reproduce the scenario as it requires installing older versions of libraries, building an app that uses the old libraries (including an older version of Vite, MUI, date-fns, and mui/x-date-pickers), setting up a page that uses AdapterDateFns (the old version), having all that cached via Vite, then upgrading all the libraries, switching to the new AdapterDateFnsV3, and trying to run the app again.  When running the new app with the latest versions of the aforementioned libraries and utilizing the AdapterDateFnsV3, Vite will fail on runtime due to a pre-bundle issue (possibly a caching issue) whereby it tries to pre-bundle an unused AdapterDateFns file rather than relying purely on the AdapterDateFnsV3 file.  This causes an error since the AdapterDateFns.js file uses the old date-fns@v2 method for importing longFormatters, which conflicts with the upgraded date-fns@v3 since that uses a different approach to importing.

Long story short, I solved the problem by simply making this change to the source code of "node_modules\@mui\x-date-pickers\AdapterDateFns\AdapterDateFns.js":

```
import { longFormatters } from 'date-fns';
// import longFormatters from 'date-fns/_lib/format/longFormatters';
```

I commented out the failing line as seen above, and replaced it with the proper way of importing the longFormatters according to the latest version of date-fns@v3.  Hope this helps.  Unfortunately, it would be too hard to reproduce due to the aforementioned "complexity" of how the error was encountered to begin with.

**Comment by LukasTy** [comment]
> I commented out the failing line as seen above, and replaced it with the proper way of importing the longFormatters according to the latest version of date-fns@v3. Hope this helps. Unfortunately, it would be too hard to reproduce due to the aforementioned "complexity" of how the error was encountered to begin with.

@harbhub Thank you for sharing more information and steps on how you triggered this issue. 👍 
Is it hard to clear Vite pre-bundle cache where this issue is appearing? 🤔 
The fix seems somewhat fragile and in an ideal world—not needed, if it's truly a cache issue. 🤞

**Comment by harbhub** [comment]
@LukasTy Unfortunately, I'm unsure about how hard it would be to properly/intentionally clear the Vite pre-bundle cache.  I just kept my slight modification to the x-date-pickers library and moved on to other tasks.  If/When I have a chance, I'll try to find an answer to your question.

**Comment by alonando** [comment]
@LukasTy @harbhub 

Running `vite --force` will clear vite cache

I was having the same issue after the upgrade. That and using the right imports worked!

**Comment by LukasTy** [comment]
@alonando Thank you for sharing information on a cleaner solution. 🙏

**Comment by mphilip9** [comment]
> @LukasTy @harbhub
> 
> Running `vite --force` will clear vite cache
> 
> I was having the same issue after the upgrade. That and using the right imports worked!

I'm still getting this error even after clearing the vite cache. But if I implement the change from @harbhub the error goes away. Which isn't the nicest solution. Has anyone else encountered the problem and come across a different fix

**Comment by LukasTy** [comment]
@mphilip9 I'm sorry you are having problems with it, but sadly, I cannot help you, without a reproduction case. 🙈 🤷

**Comment by 4t4nner** [comment]
> I fixed it by updating the Material UI library in this manner:
> 
> .../node_modules/@mui/x-date-pickers/AdapterDateFns/AdapterDateFns.js // import longFormatters from 'date-fns/_lib/format/longFormatters'; import { longFormatters } from 'date-fns';
> 
> I commented out the old import and replaced it with the proper one based on how date-fns@v3.6.0 works.
> Leaving this solution here in case it happens to help anyone else.

```typescript
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
```

```typescrypt
// package.json:
"date-fns": "^4.1.0",
```
still works. Still not works without (:

**Comment by ZakKa89** [comment]
> [@ZakKa89](https://github.com/ZakKa89) Which package version are you on? If you are on v7, then you need to use `AdapterDateFnsV3` as [mentioned here](https://v7.mui.com/x/react-date-pickers/adapters-locale/#with-date-fns). If you are already on v8, then `AdapterDateFns` should be your choice. 😉

Thanks for the swift reply. 

Regarding my package version, it's in my original message: 8.4.0 (x-date-pickers)
Yes, I'm using AdapterDateFns. I tried reproducing the issue in a codesandbox, but no luck so far. Maybe I just need to delete node_modules, package lock, and try reinstalling again. 

https://codesandbox.io/p/sandbox/33n77g <-- Works as expected here.

**Comment by ZakKa89** [comment]
> I suspect that you have several version of the `@mui/x-date-pickers` installed. Because the import in your error message in only used in our adapters that target date-fns v2. And still the file is named `AdapterDateFns.js` I suspect that it's not our v8 but an earlier version.

You are right! I just now noticed that x-date-pickers is installed as a peerDependency as well with an older version seperately 🤦 . Thanks for the help. Will report back if I get it to work.
